### ISSUE TITLE
Zoom & pan SVG output

### ISSUE THREAD

**Issue by jakearchibald** [jakearchibald/svgomg#2]
Using

* + & - zoom buttons
* Click-drag for panning
* Touch for panning
* Mousewheel for scaling (watch out for Firefox vs Chrome wheel units)
* Pinch for scaling

**Comment by jakearchibald** [comment]
Also: some IE bugs but not sure if that's just in the VM

**Comment by paperboyo** [comment]
Hello,

First of all thank you _so_ much! Your tool already is so good, I’m lost with words...

I wouldn't want to flood you with suggestions, but here are just some connected to zooming/output preview:

- I consider vector zoom to be more useful than pixel zoom when comparing SVGs. Currently the mousewheel does pixel zoom and the only way to do a vector one I have found is to zoom with ctrl + [+ OR -] which also scales the interface. I would rather have a vector zoom as a main mousewheel zoom

- when comparing original with svgo output, I often (always!) look for things that went wrong. I used to open original and optimised SVGs in diferrent tabs, zoom them in and quickly flick through the tabs. I think it would be cool if SVGOMG would temporarily show the original graphic e.g on middle mouse press (much like ImageAlpha is doing it). It would work while zoomed in, ideally

- I do not understand ‘Mousewheel for scaling in lines (Firefox with mouse connected)’ - how is it different to pixel zoom and how do I achieve it?

- in latest Firefox 36 beta on Windows 8.1, zooming with the mousewheel does not work and the demo has extra sliders around the graphic, like so:
![svgomg](https://cloud.githubusercontent.com/assets/6032869/5905920/9b99570c-a58b-11e4-92a9-cf5f15ff5fe0.PNG)

Thank you again for your work!

Regards
m.

**Comment by jakearchibald** [comment]
@paperboyo

> I consider vector zoom to be more useful than pixel zoom when comparing SVGs

In the ticket, when I talk about "pixels" I mean the unit I get from the mousewheel event. Whether scaling is vector or not is down to the OS at the moment. It looks pixelly in Chrome & Safari to me, are you seeing it elsewhere? I had a fix for it, but it reverted it last night as it killed performance. https://github.com/jakearchibald/svgomg/issues/9

> I think it would be cool if SVGOMG would temporarily show the original graphic

Does the "Show original" option not do what you're looking for?

> I do not understand ‘Mousewheel for scaling in lines (Firefox with mouse connected)’ - how is it different to pixel zoom and how do I achieve it?

Mouse wheel events give you a delta value for the amount moved. This is in pixels, lines, or pages. I've got the zoom working if the value is pixels, but not other units. The line unit is used in Firefox if you have a mouse connected rather than a trackpad. All I need to do if find a sensible pixel value to convert lines to, but I need to get hold of a mouse to test it (on the other side of the world to my desk at the moment).

>  the directions seem to be reversed when zooming with a mousewheel for me

Mac inverts the scrollwheel. Which direction feels wrong to you? Which OS do you use the most?

> in latest Firefox 36 beta on Windows 8.1, zooming with the mousewheel does not work and the demo has extra sliders around the graphic, like so:

I think these issues are unrelated. I suspect zooming isn't working due to line delta units. Here's a ticket for the scrollbars https://github.com/jakearchibald/svgomg/issues/10

**Comment by paperboyo** [comment]
> It looks pixelly in Chrome & Safari to me, are you seeing it elsewhere?

Thanks for responding! With just a mousewheel it’s pixelated in latest betas of Chrome and Opera on Win 8.1; vector in IE11 (although, some strange things are going on there, like image travelling around the canvass like crazy and running away for good, zooming should have some sane bounds here, I think:). Mousewheel does not work in latest FF beta and latest nightly at all (nightly does away with the sliders, though). Ctrl + [- OR +] gives vector zoom in all browsers, but the interface, obviously, zooms in with the content.

> Does the "Show original" option not do what you're looking for?

Exactly that, just to far away when you are zooming in, panning different parts of the image inspecting it in total concentration hoping it survived the optimisation ;-) Maybe a keyboard shortcut, if middle mouse too hard (would like to be able to toggle it quickly without looking away from the image - easier to spot inconsistencies). Not terribly important, obviously, just a thought.

> Mouse wheel events give you a delta value for the amount moved.

Understood. Meaning: totally not understood, but that's ok. I can just turn the wheel... Ah, have seen these ‘lines’ somewhere in the control panel couple of years ago... :-D

> Mac inverts the scrollwheel. Which direction feels wrong to you? Which OS do you use the most?

Yeah, Mac seem to invert a thing or two every couple of years, yeah. I use mostly Mac. And then, mostly Windows. Subjectively, (+ Adobe cross-platformly), scrolling up should zoom in, scrolling down should zoom out. If there is no way to take a hint from the OS, then I would settle on the most common, which may or not may be what I suggested.

> Here's a ticket for the scrollbars #10

Thank you! Will remember to keep issues separate. Now let me think of some that are not actually feature requests for svgo as such.

_Chapeau bas !_
m.

**Comment by albertorestifo** [comment]
I have a very sensible mouse wheel, I accidentally zoomed too far back and it was impossible to find again the SVG. I think there is the need of a limitation on the amount you can zoom back

**Comment by shshaw** [comment]
The pixel-y issue is related to `translate3d` being used instead of just `translate`. I believe Chrome is rendering the vector layer as a raster image to move it to the GPU for faster rendering, problem is it renders at the original size so scaling up makes it blur out.

If you switch the `translate3d` to `translate` in devTools, the pixelation should not occur.

**Comment by jakearchibald** [comment]
Yep, aware of that. Performance went through the floor. Also tried switching to 2d transform on a timeout, but performance still wasn't good enough. That might be different now.
